### DISCUSSION TITLE
Inner working of the menu machine

### DISCUSSION THREAD

**Discussion by suimong** [chakra-ui/zag#227] [Q&A] [answered]
Hi team,

Thanks for making such a wonderful library! In my opinion zag is taking "reusable" component to the next level, where the actual hard part of managing component states is abstracted! 
I just started playing with the menu example, and while debugging and reading the zag source code, there's something I can't wrap my head around:
1. As shown in the following screenshot, the style attribute of the "positioner" div contains all information that will place the actual menu at the write position:
![image](https://user-images.githubusercontent.com/812576/180386921-f2e47f08-370b-4778-9f97-c57f6dda984c.png)
Presumably the attributes should come from `api.positionerProps`

2. the `api` object is obtained through `menu.connect`, so I went in there, find the `positionerProps` definition as shown below:
![image](https://user-images.githubusercontent.com/812576/180389307-6fa61244-c151-4bfc-ae89-35e00215a78d.png)

3. It tells me that the style prop comes from `popperStyles.floating`, so continue the trace to `popperStyles` gets me in the function `getPlacementStyles`:
![image](https://user-images.githubusercontent.com/812576/180389793-7e65f0b4-e45f-41a7-958a-6ce6b0501224.png)

4. Here comes the problem: within `getPlacementStyles`, the floating prop only outputs an ojbect with two fields: `position` and `minWidth` (because after the first render the `measured` variable becomes `true` and line 53 will short circuit), there are no signs of css attributes like `--reference-width, --available-width, transform` etc.
![image](https://user-images.githubusercontent.com/812576/180390260-e3ff1b39-9aa5-4389-9911-d5dcc0c63630.png)
 
So my question is: how (and where) does zag put all those attributes in the actual dom element?

**Comment by anubra266** [comment, marked as answer]
@suimong https://github.com/chakra-ui/zag/blob/d786c6cc33724a486cd085415f030af0dab8b34b/packages/utilities/popper/src/get-placement.ts#L66-L76

**Comment by suimong** [comment]
Thank you @anubra266 ! So `getPlacement` is called inside `computePlacement` which is one of the `activities` of the menu machine, and when the machine is in state "open", then `computePlacement` gets called, am I right?
